modules/gapi/src/executor/gtbbstreamingexecutor.hpp

```diff
@@ -0,0 +1,168 @@
+#include <unordered_set>
+#include <vector>
+
+#if defined(HAVE_TBB)
+#  include <tbb/concurrent_queue.h>
+template<typename T> using QueueClass = tbb::concurrent_bounded_queue<T>;
+#else
+#  include "executor/conc_queue.hpp"
+template<typename T> using QueueClass = cv::gapi::own::concurrent_bounded_queue<T>;
+#endif // TBB
+
+#include "executor/gstreamingexecutor.hpp"
+
+namespace cv {
+namespace gimpl {
+
+namespace stream {
+struct Start {};
+struct Stop {
+    enum class Kind {
+        HARD, // a hard-stop: end-of-pipeline reached or stop() called
+        CNST, // a soft-stop emitted for/by constant sources (see QueueReader)
+    } kind = Kind::HARD;
+    cv::GRunArg cdata; // const data for CNST stop
+};
+struct Result {
+    cv::GRunArgs      args;  // Full results vector
+    std::vector<bool> flags; // Availability flags  (in case of desync)
+};
+using Cmd = cv::util::variant
+    < cv::util::monostate
+    , Start                // Tells emitters to start working. Not broadcasted to workers.
+    , Stop                 // Tells emitters to stop working. Broadcasted to workers.
+    , cv::GRunArg          // Workers data payload to process.
+    , Result               // Pipeline's data for gout()
+    , cv::gimpl::Exception // Exception which is thrown while execution.
+   >;
+
+// Interface over a queue. The underlying queue implementation may be
+// different. This class is mainly introduced to bring some
+// abstraction over the real queues (bounded in-order) and a
+// desynchronized data slots (see required to implement
+// cv::gapi::desync)
+class Q {
+public:
+    virtual void push(const Cmd &cmd) = 0;
+    virtual void pop(Cmd &cmd) = 0;
+    virtual bool try_pop(Cmd &cmd) = 0;
+    virtual void clear() = 0;
+    virtual ~Q() = default;
+};
+
+// A regular queue implementation
+class SyncQueue final: public Q {
+    QueueClass<Cmd> m_q;    // FIXME: OWN or WRAP??
+
+public:
+    virtual void push(const Cmd &cmd) override { m_q.push(cmd); }
+    virtual void pop(Cmd &cmd)        override { m_q.pop(cmd);  }
+    virtual bool try_pop(Cmd &cmd)    override { return m_q.try_pop(cmd); }
+    virtual void clear()              override { m_q.clear(); }
+
+    void set_capacity(std::size_t c) { m_q.set_capacity(c);}
+};
+
+// Desynchronized "queue" implementation
+// Every push overwrites value which is not yet popped
+// This container can hold 0 or 1 element
+// Special handling for Stop is implemented (FIXME: not really)
+class DesyncQueue final: public Q {
+    cv::gapi::own::last_written_value<Cmd> m_v;
+
+public:
+    virtual void push(const Cmd &cmd) override { m_v.push(cmd); }
+    virtual void pop(Cmd &cmd)        override { m_v.pop(cmd);  }
+    virtual bool try_pop(Cmd &cmd)    override { return m_v.try_pop(cmd); }
+    virtual void clear()              override { m_v.clear(); }
+};
+} // namespace stream
+
+class GTBBStreamingExecutor final : public GStreamingExecutor {
+protected:
+    const bool m_desync;
```

Comment:
I suggest to implement TWO different `GTBBStreamingExecutor`:
`GTBB_SYNC_StreamingExecutor` ---> `GTBBStreamingExecutor` <-- `GTBB_DESYNC_StreamingExecutor`

Do do not stick flag-based logic through single and huge GTBBStreamingExecutor

Reply:
@sivanov-work wouldn't it involve a huge chunk of code being copy-pasted?